src/platform_impl/windows/icon.rs

```diff
@@ -1,108 +1,162 @@
-use std::{io, mem, os::windows::ffi::OsStrExt, path::Path, ptr};
+use std::{fmt, io, iter::once, mem, os::windows::ffi::OsStrExt, path::Path, ptr, sync::Arc};
 
 use winapi::{
     ctypes::{c_int, wchar_t},
     shared::{
-        minwindef::{BYTE, LPARAM, WPARAM},
+        minwindef::{BYTE, LPARAM, WORD, WPARAM},
         windef::{HICON, HWND},
     },
+    um::libloaderapi,
     um::winuser,
 };
 
-use crate::icon::{Icon, Pixel, PIXEL_SIZE};
+pub use crate::icon::BadIcon;
+use crate::icon::*;
 
 impl Pixel {
     fn to_bgra(&mut self) {
         mem::swap(&mut self.r, &mut self.b);
     }
 }
 
+impl RgbaIcon {
+    fn into_windows_icon(self) -> Result<WinIcon, BadIcon> {
+        let mut rgba = self.rgba;
+        let pixel_count = rgba.len() / PIXEL_SIZE;
+        let mut and_mask = Vec::with_capacity(pixel_count);
+        let pixels =
+            unsafe { std::slice::from_raw_parts_mut(rgba.as_mut_ptr() as *mut Pixel, pixel_count) };
+        for pixel in pixels {
+            and_mask.push(pixel.a.wrapping_sub(std::u8::MAX)); // invert alpha channel
+            pixel.to_bgra();
+        }
+        assert_eq!(and_mask.len(), pixel_count);
+        let handle = unsafe {
+            winuser::CreateIcon(
+                ptr::null_mut(),
+                self.width as c_int,
+                self.height as c_int,
+                1,
+                (PIXEL_SIZE * 8) as BYTE,
+                and_mask.as_ptr() as *const BYTE,
+                rgba.as_ptr() as *const BYTE,
+            ) as HICON
+        };
+        if !handle.is_null() {
+            Ok(WinIcon::_new(handle))
+        } else {
+            Err(BadIcon::OsError(io::Error::last_os_error()))
+        }
+    }
+}
+
 #[derive(Debug)]
 pub enum IconType {
     Small = winuser::ICON_SMALL as isize,
     Big = winuser::ICON_BIG as isize,
 }
 
-#[derive(Clone, Debug)]
+#[derive(Debug)]
+struct RaiiIcon {
+    handle: HICON,
+}
+
+#[derive(Clone)]
 pub struct WinIcon {
-    pub handle: HICON,
+    inner: Arc<RaiiIcon>,
 }
 
 unsafe impl Send for WinIcon {}
 
 impl WinIcon {
-    #[allow(dead_code)]
-    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self, io::Error> {
-        let wide_path: Vec<u16> = path.as_ref().as_os_str().encode_wide().collect();
+    pub fn as_raw_handle(&self) -> HICON {
+        self.inner.handle
+    }
+
+    pub fn from_path<P: AsRef<Path>>(
+        path: P,
+        size: Option<(usize, usize)>,
+    ) -> Result<Self, BadIcon> {
+        let wide_path: Vec<u16> = path
+            .as_ref()
+            .as_os_str()
+            .encode_wide()
+            .chain(once(0))
+            .collect();
+        // 0 indicates that we want to use the actual width and height
+        let (width, height) = size.unwrap_or((0, 0));
```

Comment:
We want to use `LR_DEFAULTSIZE` as the default for this, not `0`. Icon files can contain multiple sizes, and if we specify `0` and the default size Windows will load a large version of the icon and scale down instead of loading the appropriate-sized icon for the system configuration.